### PR TITLE
Upgrade Apache to 2.4.53

### LINKED ISSUE #2018: Upgrade Apache to 2.4.53
Contains security fixes

### PR DESCRIPTION
**Pull request by stondino00**
Contains security fixes.

Fixes #2018

### CODE REVIEW

**Comment by stondino00** [comment]
@kyrofa not sure why this failed on me.

**Comment by pachulo** [comment]
All the integrations tests are failing, i'ts like Apache is not starting OK...

**Comment by kyrofa** [comment]
Yeah that is what it looks like. I really wish I could finish setting up publishing PR snaps from github actions. @stondino00 have you tried building this and poking at it?

**Comment by stondino00** [comment]
Hey @pachulo @kyrofa!

Sorry I've not tested a build of this.  I'm only as well versed enough to make these patches for the updated dependencies to help everyone out.  You and Pachulo are the brains here :).  What's interesting is Pachulo's 23.0.3 build finished fine it appears yesterday.  Must be something on that latest version that they changed.  I can reach out to the mailing list and see what we need to change for this latest version.

**Comment by stondino00** [comment]
Maybe try merging 23.0.3 into develop and I'll rebase this and see if that does anything.

**Comment by stondino00** [comment]
Thoughts?

**Comment by stondino00** [comment]
Again, I'm not very knowledgeable past helping with patching the dependencies.

Seems to be timing out on this sleep, wonder if it should be longer???




RuntimeError:
       Timed out trying to access Nextcloud: https://localhost
     # ./integration/spec_helper.rb:222:in `rescue in wait_for_nextcloud'
     # ./integration/spec_helper.rb:204:in `wait_for_nextcloud'
     # ./integration/spec_helper.rb:181:in `enable_https'
     # ./integration/change_port_spec.rb:30:in `block (3 levels) in <top (required)>'
     # ------------------
     # --- Caused by: ---
     # Timeout::Error:
     #   execution expired
     #   ./integration/spec_helper.rb:218:in `sleep'

**Comment by kyrofa** [comment]
Untested yet, but I placed a build for all architectures in `latest/beta/pr-2019`.

**Comment by stondino00** [comment]
> latest/beta/pr-2019

Refreshed to this PR and all worked fine.  Nextcloud doesn't show me the version of Apache that's installed like it does for mysql and php.  Did you make any changes?

**Comment by kyrofa** [comment]
No, I just pulled this branch and built it.

**Comment by stondino00** [comment]
> No, I just pulled this branch and built it.

Interesting

**Comment by stondino00** [comment]
I've been running this PR since yesterday and no issues.  

image

**Comment by pachulo** [comment]
> Nextcloud doesn't show me the version of Apache that's installed like it does for mysql and php. Did you make any changes?

This is the expected behaviour. Were you seeing it before upgrading to this branch?

**Comment by stondino00** [comment]
No I know it doesn't show the version.  It's helpful to see that on the other dependencies so you know your running the current version of them when testing these PR's.  Any thoughts on how to move forward with this?

**Comment by pachulo** [comment]
> Any thoughts on how to move forward with this?

I'm sorry, I don't currently have much free time for this. But I guess that trying to execute the tests locally with this branch should shred some light.

**Comment by kyrofa** [comment]
I'm actually having trouble getting this to fail locally. I'm not sure what's happening, here.

**Comment by kyrofa** [comment]
Okay I finally managed to pull a log from the runner:

> 2022-04-14T02:35:09Z nextcloud.apache[20801]: httpd: error while loading shared libraries: libpcre2-8.so.0: cannot open shared object file: No such file or directory

No idea why we're not seeing that when running locally.

**Comment by kyrofa** [comment]
Probably because it doesn't seem to NEED libprce2-8:

```
$ ldd httpd
	[...]
	libpcre.so.3 => /lib/x86_64-linux-gnu/libpcre.so.3 (0x00007fd10e593000)
	[...]
```

I would start to wonder if this was an issue with the github runner itself, but these tests pass on other PRs. I'm very confused by this.

**Comment by kyrofa** [comment]
Alright I think I've got a handle on this.

**Comment by kyrofa** [comment]
Alright, got it. We had a libpcre2 dependency leaking in from the build host, when libpcre3 is contained in the core18 snap. The proper solution is a more isolated build (see #2037). I was confused why this was the issue though-- why did previous apache builds work fine? Well, I took a closer look at the [changelog](https://downloads.apache.org/httpd/CHANGES_2.4.53) for this apache release:

> * Support pcre2 (10.x) library in place of the now end-of-life pcre (8.x)
     for regular expression evaluation. This depends on locating pcre2-config.
     [William Rowe, Petr Pisar <ppisar redhat.com>, Rainer Jung]

Initially I figured that was fine, no problem. "Who cares about pcre2," I thought, "we're using pcre3!" But... that error. I took a closer look at the libpcre packages. Check this out:

```
$ apt-cache search libpcre | grep dev
libpcre2-dev - New Perl Compatible Regular Expression Library - development files
libpcre3-dev - Old Perl 5 Compatible Regular Expression Library - development files
```

Uh. So libpcre3 is "Old", and libpcre2 is "New"? No way. Indeed:

```
$ apt-cache show libpcre3-dev
# [...]
Description-en: Old Perl 5 Compatible Regular Expression Library - development files
 This is a library of functions to support regular expressions whose syntax
 and semantics are as close as possible to those of the Perl 5 language.
 .
 New packages should use the newer pcre2 packages, and existing
 packages should migrate to pcre2.
```

What on earth. I found more details in [this thread](https://lists.apache.org/thread/nokzm0yf1d82qf3c9b41cj72y16bjgos), summarized here:

> Hello team,
Is there any plan to migrate from pcre3 to pcre2 ?
[...]
TL;DR> pcre3 is obsolete and upstream don't want to fix it any more. I 
propose a MBF to track our progress in getting rid of it for Bookworm
>
> PCRE is the perl-compatible regular expression library, 
https://pcre.org// For historical reasons, the old PCRE library ended up 
as libpcre3 in Debian. I'm going to call it pcre3 hereafter.
>
> In 2015, PCRE upstream brought out pcre2, re-architectured to be be more 
secure and easier to maintain (it's been in Debian since stretch). Many 
large (and small) projects have successfully migrated to pcre2.
[...]

So basically, previous versions of Apache didn't support the version of libpcre on the github action runner, so it was ignored. This version added support for it, so it was picked up, but only at build time-- not shipped in the snap. Which caused it to bail at runtime.

> The proper solution is a more isolated build (see #2037).
>
> ~ me, 30 seconds ago

I'm not wrong, that IS a solution and it's probably necessary, but now that I've learned more about the libpcre versioning wonkiness, we should also be using the newer version. So the PROPER proper solution is to do both. The former is done in #2037, the latter is done here.

**Comment by stondino00** [comment]
Good work!

**Review by kyrofa — APPROVED** [review]
At long last, +1. Sorry @stondino00, this one was really rough to track down! Thanks as always.

**Comment by pachulo** [comment]
Wow, congrats & thanks @kyrofa !! This was a nasty one to figure out...
